pytorch_lightning/metrics/functional/image_gradients.py

```diff
@@ -0,0 +1,74 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Tuple
+
+import torch
+
+
+def _image_gradients_validate(img: torch.Tensor) -> torch.Tensor:
+    """ Validates whether img is a 4D torch Tensor """
+
+    if not isinstance(img, torch.Tensor):
+        raise TypeError(f"The `img` expects a value of <torch.Tensor> type but got {type(img)}")
+    if img.ndim != 4:
+        raise RuntimeError(f"The `img` expects a 4D tensor but got {img.ndim}D tensor")
+
+
+def _compute_image_gradients(img: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    """ Computes image gradients (dy/dx) for a given image """
+
+    batch_size, channels, height, width = img.shape
+
+    dy = img[..., 1:, :] - img[..., :-1, :]
```

Comment:
Everything is looking good to me, but I've just one question. Gradient of an image is a standard computer vision task, and usually it's implemented with convolutions and kernels (the implemented version is the simplest one, with kernel [-1, +1], but there are more advance operators e.g., sobel [wiki](https://en.wikipedia.org/wiki/Sobel_operator)). Have you thought about the computational cost of accessing pytorch tensor directly instead of using convolution operators? A simple test with `timeit` can show the difference in terms of computational time. You can also use the classical approach with convs and kernels as a unittest! Just a suggestion btw, we can discuss this :)

Reply:
Hey @gianscarpe, thanks for pointing this out! The reason behind sticking to the 1-step finite difference method is as follows, 
1. The original idea was to make "depth-loss" available out of the box. In most of the papers that use depth-loss, I have observed they stick to the finite-difference method (case in point the [densedepth](https://github.com/ialhashim/DenseDepth/blob/master/Tensorflow/loss.py) paper)
2. TF follows the 1-step finite-difference out of box. 

But I understand that making sobel etc available is also equally important, for certain use-cases may explicitly need sobel etc. So like I guess the natural solution is making available other advanced filters with 1-step as the default, and letting the user change if they feel the need to do so. Once I have completed the original goal of including depth-loss in lightning, I would love to work on this and make available more advanced filters if thats okay!

Reply:
I propose that we merge the PR as it is now, and then we can always add `method` argument in a follow up PR that determines how the difference should be computed.

Reply:
Hi @jaypmorgan, I see your point! I'm not sure, mine is just an empirical consideration. tf uses channel-last tensors, while pytorch uses channel-first. I noticed some slowdowns in my own code when reshaping channel-first tensors, due to the computational overhead. 
Do you mind check with `timeit` if I'm wrong? A simple conv-based implemetation for this is the following:
```
>> image = torch.arange(0, 1*1*5*5, dtype=torch.float32)
>> image = torch.reshape(image, (1, 1, 5, 5))
>>kernel = torch.tensor([[[[-1, +1]]]]).float()
>> dx = F.conv1d(input=image, weight=k)
tensor([[[[1., 1., 1., 1.],
          [1., 1., 1., 1.],
          [1., 1., 1., 1.],
          [1., 1., 1., 1.],
          [1., 1., 1., 1.]]]])

>>> dy = F.conv1d(image, k.transpose(-1, -2)) # conv1d with transposed kernel
tensor([[[[5., 5., 5., 5., 5.],
          [5., 5., 5., 5., 5.],
          [5., 5., 5., 5., 5.],
          [5., 5., 5., 5., 5.]]]])
```

Reply:
Hey @gianscarpe, I ran the code snippet above against the implemented 1-step difference using timeit, and the test code snippets with the results are given bellow:

Hardware: HP Pavillion 15, Intel i5 8th gen 

Using convolutions (code snippet given above):

```python
import timeit

setup = '''
import torch
import torch.nn.functional as F
from pytorch_lightning.metrics.functional.image_gradients import image_gradients

BATCH_SIZE = 5
HEIGHT = 5
WIDTH = 5
CHANNELS = 1

single_channel_img = torch.arange(0, 1 * HEIGHT * WIDTH * CHANNELS, dtype=torch.float32)
single_channel_img = torch.reshape(single_channel_img, (CHANNELS, HEIGHT, WIDTH))
image = torch.stack([single_channel_img for _ in range(BATCH_SIZE)], dim=0)

kernel = torch.tensor([[[[-1, +1]]]]).float()
'''

code = '''dx = F.conv1d(input=image, weight=kernel)
dy = F.conv1d(image, kernel.transpose(-1, -2))
'''

print(timeit.timeit(setup=setup, stmt=code))
```
*Total run time:*  167.21782171400264

Using the 1-step finite difference method (as implemented in this PR):

```python

import timeit

setup = '''
import torch
from pytorch_lightning.metrics.functional.image_gradients import image_gradients

BATCH_SIZE = 5
HEIGHT = 5
WIDTH = 5
CHANNELS = 1

single_channel_img = torch.arange(0, 1 * HEIGHT * WIDTH * CHANNELS, dtype=torch.float32)
single_channel_img = torch.reshape(single_channel_img, (CHANNELS, HEIGHT, WIDTH))
image = torch.stack([single_channel_img for _ in range(BATCH_SIZE)], dim=0)
'''

code = "image_gradients(image)"

print(timeit.timeit(setup=setup, stmt=code))
```
*Total Run Time:* 46.073826758991345

I guess this difference is primarily because the 1-step diff is a very simple way to calculate gradients. And probably that compensates for the overhead during the reshaping at the end. But I guess the next natural step should be to make kernels like sobel etc available for gradient calculation!

Reply:
Nice! :) then keep it for a future issue!

Reply:
Sure!